Added get_versions method in NodeJS api with the basic test.

adismort14 opened the pull request:
This Fixes #23430 .

Details:

1. Added get_versions() method definition into bindings (core_wrap.hpp)
2. Implemented this method in bindings (core_wrap.cpp)
3. Add interface implementation to addon.ts
4. Added basic js test for this method in `src/bindings/js/node/tests/basic.test.js`

CC: @vishniakov-nikolai

vishniakov-nikolai commented on src/bindings/js/node/lib/addon.ts:
```diff
@@ -38,6 +38,7 @@ interface Core {
   readModelSync(modelBuffer: Uint8Array, weightsBuffer?: Uint8Array): Model;
   importModelSync(modelStream: Buffer, device: string): CompiledModel;
   getAvailableDevices(): string[];
+  getVersions():{ deviceName: string; version: string }[];
```
Misalignment with bindings signature

adismort14 replied:
Fixed this in the latest commit. Kindly review it.

vishniakov-nikolai commented on src/bindings/js/node/src/core_wrap.cpp:
```diff
@@ -232,6 +233,25 @@ Napi::Value CoreWrap::get_available_devices(const Napi::CallbackInfo& info) {
 }
 
+Napi::Value CoreWrap::get_versions(const Napi::CallbackInfo& info) {
+    const auto& devices_map = _core.get_versions(info[0].As<Napi::String>().Utf8Value());
```
What if not string type would be passed or args would absent?

adismort14 replied:
I have added the required checks for this.

vishniakov-nikolai commented on src/bindings/js/node/src/core_wrap.cpp:
```diff
@@ -232,6 +233,25 @@ Napi::Value CoreWrap::get_available_devices(const Napi::CallbackInfo& info) {
 
+Napi::Value CoreWrap::get_versions(const Napi::CallbackInfo& info) {
+    const auto& devices_map = _core.get_versions(info[0].As<Napi::String>().Utf8Value());
+    Napi::Array js_devices = Napi::Array::New(info.Env(), devices_map.size());
```
Why array should have size = passed string length?

adismort14 replied:
`_core.get_versions` returns a map containing the device and the deviceVersion. Thus the length of js_devices array should be the same as that of devices_map.
I have referred to this doc for the same. 
```std::map<std::string, ov::Version> ov::Core::get_versions(const std::string &device_name) const
Returns device plugins version information. Device name can be complex and identify multiple devices at once like HETERO:CPU,GPU; in this case, std::map contains multiple entries, each per device.

Parameters:
device_name – Device name to identify a plugin.

Returns:
A vector of versions.
```

vishniakov-nikolai replied:
My bad, I wrongly interpreted map type.

vishniakov-nikolai commented on src/bindings/js/node/src/core_wrap.cpp:
```diff
@@ -232,6 +233,25 @@ Napi::Value CoreWrap::get_available_devices(const Napi::CallbackInfo& info) {
+        js_device.Set("deviceName", Napi::String::New(info.Env(), dev.first));
+        std::string versionStr = dev.second.buildNumber;
+        versionStr += " ";
+        versionStr += dev.second.description;
```
I propose to use structure, do not concatenate parameters in string

vishniakov-nikolai commented on src/bindings/js/node/tests/basic.test.js:
```diff
@@ -20,6 +20,13 @@ it('Core.getAvailableDevices()', () => {
+  const deviceVersion = core.getVersions(devices[0]);
+
+  assert.ok(deviceVersion.size()>=1);
+});
```
Tests have to check behavior in the cases when incorrect parameters passed, parameters are absent and that result has expected format.

vishniakov-nikolai reviewed (changes requested):
Please, pay attention on comments. If you need more information, feel free to ask.

adismort14 commented:
Here are the updates from my end post the review:

1. [Review Comment](https://github.com/openvinotoolkit/openvino/pull/23648#discussion_r1539573589) Fixed the binding signatures.
2. [Review Comment](https://github.com/openvinotoolkit/openvino/pull/23648#discussion_r1539575566) Added the check for no args passed and non string convertible arguments passed.
3. [Review Comment](https://github.com/openvinotoolkit/openvino/pull/23648#discussion_r1539604579) Modified the code to contain the parameters in an object of the form:
```js
{
    deviceName: String,
    version: String,
    versionDescription: String
}
```
4. [Review Comment](https://github.com/openvinotoolkit/openvino/pull/23648#discussion_r1539625384) Added more tests to test the unexpected behaviour. I have a question regarding this: How do I run the tests locally to check whether the tests I have written work as expected and not have to discover them during the github workflow checks? Could you share any doc or share some pointers for the same?

@vishniakov-nikolai  Kindly review them and let me know if any changes are required.
Thanks!

vishniakov-nikolai commented on src/bindings/js/node/lib/addon.ts:
```diff
@@ -38,6 +38,7 @@ interface Core {
   readModelSync(modelBuffer: Uint8Array, weightsBuffer?: Uint8Array): Model;
   importModelSync(modelStream: Buffer, device: string): CompiledModel;
   getAvailableDevices(): string[];
+  getVersions(deviceName: string):{ deviceName: string, version: string, versionDescription: string }[];
```
I propose to follow original format: 
```suggestion
  getVersions(deviceName: string): { 
    [deviceName: string]: { 
      buildNumber: string,
      description: string,
    },
  };
```

vishniakov-nikolai replied:
Also, please check that your IDE is using and following project eslint rules

vishniakov-nikolai commented on src/bindings/js/node/src/core_wrap.cpp:
```diff
@@ -232,6 +233,40 @@ Napi::Value CoreWrap::get_available_devices(const Napi::CallbackInfo& info) {
 
+Napi::Value CoreWrap::get_versions(const Napi::CallbackInfo& info) {
+    // If no arguments are passed into the method.
+    if(info.Length()==0) {
```
Please, make sure that your code formatted using openvino [clang format](https://github.com/openvinotoolkit/openvino/blob/master/src/.clang-format).

vishniakov-nikolai commented on src/bindings/js/node/src/core_wrap.cpp:
```diff
@@ -232,6 +233,40 @@ Napi::Value CoreWrap::get_available_devices(const Napi::CallbackInfo& info) {
+        //   deviceName: String,
+        //   version: String,
+        //   versionDescription: String
+        // }
```
Comment like this is extra. Code should be selfdocumented.

adismort14 commented:
Hey!
I have modified the code based on your review.

1. [Review Comment](https://github.com/openvinotoolkit/openvino/pull/23648#discussion_r1543177538) I have modified the binding signature according to this as well as the get_versions function code in `core_wrap.cpp`.
2. [Review Comment](https://github.com/openvinotoolkit/openvino/pull/23648#discussion_r1543183489) Formatted the code using clang format.
3. I have also corrected my tests based on the modifications. I have run them locally and they are passing. Removed all the unwanted comments and eslint changes.

@vishniakov-nikolai Please do review the changes when you get the time. :smile:

vishniakov-nikolai commented on src/bindings/js/node/src/core_wrap.cpp:
```diff
@@ -232,6 +233,31 @@ Napi::Value CoreWrap::get_available_devices(const Napi::CallbackInfo& info) {
 
+Napi::Value CoreWrap::get_versions(const Napi::CallbackInfo& info) {
+    if (info.Length() == 0) {
+        reportError(info.Env(), "No argument provided in the getVersions() method call.");
```
Include useful info into error message
```suggestion
        reportError(info.Env(), "get_versions expects 1 argument of string type");
```

vishniakov-nikolai commented on src/bindings/js/node/src/core_wrap.cpp:
```diff
@@ -232,6 +233,31 @@ Napi::Value CoreWrap::get_available_devices(const Napi::CallbackInfo& info) {
+        return info.Env().Undefined();
+    }
+    const auto& devices_map = _core.get_versions(device_arg.ToString());
+    Napi::Object versionsObject = Napi::Object::New(info.Env());
```
Better to use snake case notation at c++ side
```suggestion
    Napi::Object versions_object = Napi::Object::New(info.Env());
```

vishniakov-nikolai commented on src/bindings/js/node/tests/basic.test.js:
```diff
@@ -20,6 +20,31 @@ it('Core.getAvailableDevices()', () => {
+describe('Core.getVersions()', () => {
+  const devices = core.getAvailableDevices();
+
+  it('getVersions(validDeviceName: string)', () => {    
+    const deviceVersion = core.getVersions(devices[0]);
+    assert.ok(Object.keys(deviceVersion).length >= 1);
+  });
```
I don't see any purpose to check it in this way. 
My proposals:
- check that `getVersions` returns object
- check that object contains property in specified format

vishniakov-nikolai reviewed (commented):
Almost ready, minor comments

vishniakov-nikolai commented on src/bindings/js/node/tests/basic.test.js:
```diff
@@ -20,6 +20,34 @@ it('Core.getAvailableDevices()', () => {
+
+  it('getVersions() throws if non string coercable arguments are passed into the function', () => {
+    assert.throws(
+      () => core.getVersions({deviceName:devices[0]}),
```
```suggestion
      () => core.getVersions({ deviceName: devices[0] }),
```

vishniakov-nikolai commented:
@adismort14 please, merge latest master to avoid conflicts

adismort14 commented on src/bindings/js/node/tests/basic.test.js:
```diff
@@ -20,6 +20,34 @@ it('Core.getAvailableDevices()', () => {
+  it('getVersions(validDeviceName: string)', () => {    
+    const deviceVersion = core.getVersions(devices[0]);
+    assert.strictEqual(typeof deviceVersion, 'object');
+    assert.strictEqual(typeof deviceVersion.CPU, 'object');
```
I have hardcoded the value to 'CPU' in this [commit](https://github.com/openvinotoolkit/openvino/pull/23648/commits/ad2dbcfbb189c7331f32f6832a3e48fd3b7c2a09).

adismort14 commented:
Do share your review based on the latest commit. Have merged and resolved the merge conflicts. @vishniakov-nikolai
